File: Counter_Sliding_Window_11_Nov_2023/0084_largest_rectange_in_histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        heights.append(0)
        n = len(heights)
        s = [-1]
        area = 0
        for i in range(n):
            while heights[s[-1]] > heights[i]:
                h = heights[s.pop()]
                w = i - s[-1] - 1
                area = max(area, h * w)
            s.append(i)
        return area

File: Counter_Sliding_Window_11_Nov_2023/0084-largest-rectange-in-histogram.go
package main

func largestRectangleArea(heights []int) int {
    heights = append(heights, 0)
    n := len(heights)
    s := []int{-1}
    ans := 0

    for i := 0; i < n; i++ {
        for len(s) > 1 && heights[i] < heights[s[len(s)-1]] {
            h := heights[s[len(s)-1]]
            s = s[:len(s)-1]
            w := i - s[len(s)-1] - 1
            ans = max(ans, h * w)
        }
        s = append(s, i)
    }
    return ans
}
